### PR TITLE
Add hash_dataset and make dataframe payload compression functions public

### PR DESCRIPTION
**Pull request by fjetter**
# Description:

This should allow for partition or group based hashing of the data and should make things easier if we need to verify data integrity of some sorts.

I also moved the pack/unpack logic to compress the payload into a public package. I think this can be quite useful outside of "standard" kartothek workflows as long as it's not part of dask, yet

### CODE REVIEW

File: kartothek/io/dask/dataframe.py [hunk 1]
```diff
@@ -405,3 +406,61 @@ def collect_dataset_metadata(
...
+    table: str = SINGLE_TABLE,
+    predicates: Optional[PredicatesType] = None,
+    factory: Optional[DatasetFactory] = None,
+):
+    """
+    Calculate a partition wise, or group wise, hash of the dataset.
+
+    .. note::
+
+        We do not guarantee the hash values to remain constant accross versions.
```
Comment (lr4d) [inline kartothek/io/dask/dataframe.py:435]: What about same values but different data types (e.g. uint64(1) == int64(1))?

File: kartothek/io/dask/dataframe.py [hunk 2]
```diff
@@ -405,3 +406,61 @@ def collect_dataset_metadata(
...
     return ddf
+
+
+def _unpack_hash(df, unpack_meta):
+    df2 = unpack_payload_pandas(df, unpack_meta)
+    return _hash_partition(df2)
+
+
+def _hash_partition(part):
+    return pd.util.hash_pandas_object(part, index=False).sum().astype("uint64")
```
Comment (lr4d) [inline kartothek/io/dask/dataframe.py:417]: I suppose the purpose of `uint64` is to prevent overflows here
```suggestion
    return pd.util.hash_pandas_object(part, index=False).astype("uint64").sum()
```

Comment (fjetter) [reply kartothek/io/dask/dataframe.py:417]: No this is indeed expected. Summing the uint will cause overflows but these are handled gracefully. The result is then an int64 which may be negative. to stick with the practice of having uints (as is the case for the original pandas hash) I cast them back (which works also for negative numbers)


```python
In [7]: pd.Series([-1]).astype("uint64")
Out[7]:
0    18446744073709551615
dtype: uint64
```

File: kartothek/io/dask/dataframe.py (hunk 1 above)
Comment (fjetter) [reply kartothek/io/dask/dataframe.py:435]: can you rephrase the question, please? Are you referring to a guarantee about type stability or what are you referring to here? This remark shall simply indicate that we may (or may not) change the implementation without notice and we do not necessarily recommend to use this implementation for long term storage or similar.

For instance, we may want to change how hashes are combined, e.g. instead of doing `hash(partition).sum()` we may want to go for another hashing of the values, e.g. `hash(*hash(partition))` and I don't want anybody to rely on this implementation detail

Comment (lr4d) [reply kartothek/io/dask/dataframe.py:435]: Sorry for the lack of clarity. I meant to say that we might note here whether 2 datasets with the same values but different dtypes will have the same hash.

File: kartothek/io/dask/dataframe.py (hunk 2 above)
Comment (lr4d) [reply kartothek/io/dask/dataframe.py:417]: Makes sense. Just wondering (not suggesting), could we _prevent_ overflows here?

Comment (fjetter) [reply kartothek/io/dask/dataframe.py]: only if we'd use a hash function with a different domain. the pandas hash maps to the entire uint64 space, i.e. as soon as we apply a sum we're in overflow territory. Everything else would need it's dedicated implementation mapping to a smaller domain

File: kartothek/io/dask/dataframe.py (hunk 1 above)
Comment (fjetter) [reply kartothek/io/dask/dataframe.py:437]: That's also quite another topic entirely. For the sake of the argument, though

TL;DR

Same value == same hash

different type family -> different value -> different hash

```python
In [1]: import pandas as pd

In [2]: pd.util.hash_pandas_object(pd.DataFrame({"int": [1]})) == pd.util.hash_pandas_object(pd.DataFrame({"int": [1]}).astype({"int": "int8"}))
Out[2]:
0    True
dtype: bool

In [3]: pd.util.hash_pandas_object(pd.DataFrame({"int": [1]})) == pd.util.hash_pandas_object(pd.DataFrame({"int": [1]}).astype({"int": "int32"}))
Out[3]:
0    True
dtype: bool

In [4]: pd.util.hash_pandas_object(pd.DataFrame({"int": [1]})) == pd.util.hash_pandas_object(pd.DataFrame({"int": [1.0]}))
Out[4]:
0    False
dtype: bool
```

File: kartothek/io/dask/dataframe.py (hunk 2 above)
Comment (lr4d) [reply kartothek/io/dask/dataframe.py]: Even if we'd put it into `np.complex128` someway (for lack of `uint128`) we could still overflow when `len(part) >= 2^64 ~= 1.8^19`

File: kartothek/io/dask/dataframe.py (hunk 1 above)
Comment (lr4d) [reply kartothek/io/dask/dataframe.py:437]: Thanks for the explanation.
```suggestion
        We do not guarantee the hash values to remain constant accross versions.
        It should not be expected that the hash of the same value in different data types is identical.
```

File: kartothek/io/dask/dataframe.py [hunk 3]
```diff
@@ -405,3 +406,65 @@ def collect_dataset_metadata(
...
+@default_docs
+def hash_dataset(
+    store: Optional[Callable[[], KeyValueStore]] = None,
+    dataset_uuid: Optional[str] = None,
+    subset=None,
+    group_key=None,
+    table: str = SINGLE_TABLE,
+    predicates: Optional[PredicatesType] = None,
+    factory: Optional[DatasetFactory] = None,
+):
```
Comment (lr4d) [inline kartothek/io/dask/dataframe.py:431]: Please add the return type

Comment (lr4d) [reply kartothek/io/dask/dataframe.py:431]: And since it's a ddf, would you mind describing it in the docstring also?
